collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt

```diff
@@ -0,0 +1,242 @@
+
+        set(value) {
+            field = value
+            generalSettings.save("formChooserListSortingOrder", value)
+            sortAndFilter()
+        }
+
+    var filterText: String = ""
+        set(value) {
+            field = value
+            sortAndFilter()
+        }
+
+    private val shouldHideOldFormVersions: Boolean
+        get() {
+            return generalSettings.getBoolean(ProjectKeys.KEY_HIDE_OLD_FORM_VERSIONS)
+        }
+
+    init {
+        loadFromDatabase()
+        syncWithStorage()
+    }
+
+    private fun loadFromDatabase() {
+        isFormLoadingRunning.value = true
+        viewModelScope.launch(coroutineDispatchersProvider.io) {
+            var newListOfForms = formsRepository
+                .all
+                .filter {
+                    !it.isDeleted
+                }.map { form ->
+                    BlankFormListItem(
+                        databaseId = form.dbId,
+                        formId = form.formId,
+                        formName = form.displayName,
+                        formVersion = form.version ?: "",
+                        geometryPath = form.geometryXpath ?: "",
+                        dateOfCreation = form.date,
+                        dateOfLastUsage = 0,
+                        contentUri = FormsContract.getUri(projectId, form.dbId)
+                    )
+                }
+
+            if (shouldHideOldFormVersions) {
+                newListOfForms = newListOfForms.groupBy {
+                    it.formId
+                }.map { (_, itemsWithSameId) ->
+                    itemsWithSameId.sortedBy {
+                        it.formVersion
+                    }.last()
+                }
+            }
+
+            withContext(coroutineDispatchersProvider.main) {
+                _allForms.value = newListOfForms.toList()
+                sortAndFilter()
+                isFormLoadingRunning.value = false
+            }
+        }
+    }
+
+    private fun syncWithStorage() {
+        changeLockProvider.getFormLock(projectId).withLock { acquiredLock ->
+            if (acquiredLock) {
+                isSyncingWithStorageRunning.value = true
+                viewModelScope.launch(coroutineDispatchersProvider.io) {
+                    val result = formsDirDiskFormsSynchronizer.synchronizeAndReturnError()
+                    withContext(coroutineDispatchersProvider.main) {
+                        loadFromDatabase()
+                        _syncResult.value = Consumable(result)
+                        isSyncingWithStorageRunning.value = false
+                    }
+                }
+            }
+        }
+    }
+
+    fun syncWithServer(): LiveData<Boolean> {
+        logManualSyncWithServer()
+        val result = MutableLiveData<Boolean>()
+        scheduler.immediate(
+            { formsUpdater.matchFormsWithServer(projectId) },
+            { value: Boolean ->
+                loadFromDatabase()
```

Comment:
In a perfect world after syncing with server or storage if our database is update the viewmodel should be informed about this fact (repository should return livedata) and reload the forms automatically. Unfortunately we do not support returning livedata from our repositories. We should change it in the future implementing it somehow for sqlie or switching to Room where it works out of the box.

Reply:
Another approach might be to have a reactive layer in front of the repositories. We are already doing this for Instances with `InstancesAppState`. I've only started running through this PR now (and probably won't get through it all). We can discuss more later, but just wanted to throw that idea in quickly while it's in my head!